Rework rights RPCs

adonagy opened the pull request:
This PR reimplements the rights RPC and disable the routing to the protocol to serve them faster. It utilizes new storages for data needed to do the required computations.

Before merge:

- [x] look into possible leftover TODOs
- [x] cleanup debug messages (println, crit!, ...)
- [x] rustfmt...
- [x] Additionally this test has to pass http://65.21.165.82/tezedge/tezedge/16

tizoc commented on rpc/src/services/protocol/proto_001/helpers.rs:
```diff
@@ -68,168 +58,103 @@ impl RightsConstants {
+            serde_json::from_str::<serde_json::Value>(&context_proto_param.constants_data)?
+                .as_object()
+        {
+            let blocks_per_cycle = get_constant_i64(constatns_deserialized, "blocks_per_cycle")?;
```
@adonagy my idea was that for these JSONs there would just be a struct defined that would use a `#[derive(Deserialize)]` annotation to handle the parsing. Is there any reason it needs to be done manually?

adonagy replied:
That's a really good point. We should definitely use defined structs.

tizoc commented on rpc/src/services/protocol/proto_001/helpers.rs:
```diff
@@ -546,24 +463,23 @@ pub type TezosPRNGResult = Result<(i32, RandomSeedState), TezosPRNGError>;
 
     // the position of the block in its cycle; has to be i32
-    let cycle_position: i32 = level_position(level, blocks_per_cycle)?;
+    // let cycle_position: i32 = level_position(level, blocks_per_cycle)?;
```
Not used

tizoc commented on rpc/src/services/protocol/proto_001/helpers.rs:
```diff
@@ -617,3 +533,88 @@ pub fn get_prng_number(state: RandomSeedState, bound: i32) -> TezosPRNGResult {
+        requested_level: i32,
+    ) -> Result<Self, RightsConstantError> {
+        Ok(Self {
+            // (cycle_from_level(requested_level, blocks_per_cycle)?, level_position(requested_level, blocks_per_cycle)?)
```
Not used

tizoc reviewed (changes requested):
@adonagy ok, just did a first pass, I need to go over this again (and also switch to this branch locally so that I can diff between protocol versions). Other than what I commented already I see there are still a bunch of TODOs, prints and commented out code left.

adonagy commented on rpc/src/services/dev_services.rs:
```diff
@@ -164,25 +166,57 @@ pub(crate) fn get_cycle_length_for_block(
-            slog::warn!(log, "Cycle length missing"; "block" => block_hash.to_base58_check());
-            4096
-        }))
+    // TODO @adonagy: FIX THIS
```
Yes, almost done with this one

adonagy commented on rpc/src/services/protocol/mod.rs:
```diff
@@ -820,36 +949,98 @@ pub(crate) fn get_context_protocol_params(
+    metadata: &BlockMetadata,
+    key: &'static str,
+) -> Result<(i32, i32), MetadataParsingError> {
+    if let Some(level_info) = metadata.get(key) {
```
Same stuff that you were describing in the constants with the manual deserialization. We should define the metadata struct for each proto

adonagy commented on rpc/src/services/protocol/mod.rs:
```diff
@@ -820,36 +949,98 @@ pub(crate) fn get_context_protocol_params(
-            return Err(ContextParamsError::NoConstantsForBlock(
-                block_hash.to_base58_check(),
-            ));
+    slog::crit!(
```
Not needed at all, will remove.

akoptelov reviewed (commented):
There is a question on cycle position calculation, IMHO it could make it bit more clear. Also some optional suggestions on more compact code (this is not always the case, when it comes to `Option` and `Result` combination, but still it is worth a try).

And +1 to what @tizoc wrote about reusing unchanged code from previous protocols. Having 10+ *almost* similar versions is error-prone.

adonagy commented on rpc/src/services/protocol/mod.rs:
```diff
@@ -820,36 +949,98 @@ pub(crate) fn get_context_protocol_params(
-    {
-        let context = env.tezedge_context();
-        let context_hash = block_header.header.context();
+    slog::crit!(env.log(), "Getting block meta from storage");
```
Used it only for debug purposes, removed now :)

adonagy commented on rpc/src/services/protocol/proto_002/helpers.rs:
```diff
@@ -617,3 +533,88 @@ pub fn get_prng_number(state: RandomSeedState, bound: i32) -> TezosPRNGResult {
+            key: "blocks_per_cycle",
+        });
+    }
+    let cycle_position = (level % blocks_per_cycle) - 1;
```
The genesis is not included in any cycles. That means, the block's cycle_position on level 1 is 0

adonagy commented on storage/src/constants_storage.rs:
```diff
@@ -0,0 +1,78 @@
+    fn put(&self, key: &ConstantsKey, data: &str) -> Result<(), StorageError> {
+        self.kv
+            .put(key, &data.to_string())
+            .map_err(StorageError::from)
```
Yes, I assume you are right, I was trying to keep the consistency with other storages, so I did not put any thought on this.

adonagy commented on storage/src/lib.rs:
```diff
@@ -310,6 +327,33 @@ pub fn store_applied_block_result(
 
+    // populate cycle data if is present in the response
+    for cycle_data in block_result.cycle_rolls_owner_snapshots.into_iter() {
+        println!("Storing data for cycle: {}", cycle_data.cycle);
```
Removed

adonagy commented:
@tizoc @akoptelov Thanks for the review. I addressed most of the comments, there are still some small things left to be done. Will address them later today. 

As for the code duplication. I'm aware that this approach is error prone, when we have to go back and change the historic protocols. If we would merge the protocols with no changes to them, IMO it would create confusion. Let's keep it the way it is for now, and we could expand on this in the future.

bkontur commented on rpc/src/services/dev_services.rs:
```diff
@@ -162,27 +165,81 @@ pub(crate) fn get_stats_memory_protocol_runners() -> MemoryStatsResult<Vec<Memor
-        }))
+    // get the protocol hash
+    let protocol_hash =
+        match BlockMetaStorage::new(env.persistent_storage()).get_additional_data(block_hash)? {
```
@adonagy 
you can just use here cached version: `base_services::get_additional_data_or_fail`

bkontur commented on rpc/src/services/dev_services.rs:
```diff
@@ -162,27 +165,81 @@ pub(crate) fn get_stats_memory_protocol_runners() -> MemoryStatsResult<Vec<Memor
+            }
+        };
+
+    let block_level = match BlockStorage::new(env.persistent_storage()).get(block_hash)? {
```
@adonagy 
and also here: `base_services::get_raw_block_header_with_hash`,

bkontur commented on rpc/src/services/protocol/mod.rs:
```diff
@@ -820,30 +908,24 @@ pub(crate) fn get_context_protocol_params(
-            ));
-        }
+    let protocol_hash =
+        match BlockMetaStorage::new(env.persistent_storage()).get_additional_data(block_hash)? {
```
@adonagy 
also here you can use cached version: `base_services::get_additional_data_or_fail`,
after bootstrapped, this call is warmed up, so this will be even much faster

bkontur commented on rpc/src/server/router.rs:
```diff
@@ -236,16 +246,6 @@ pub(crate) fn create_routes(tezedge_is_enabled: bool) -> PathTree<MethodHandler>
             "/chains/:chain_id/blocks/:block_id/context/constants",
             protocol_handler::context_constants,
         );
-        routes.handle(
```
@adonagy @tizoc 
I see, we have constants for 009/010 finally, so I guess this 
```
routes.handle(
            hash_set![Method::GET],
            "/chains/:chain_id/blocks/:block_id/context/constants",
            protocol_handler::context_constants,
        );
```
can go outside of the if and can use rust implementation instead of protocol router

adonagy replied:
Let's keep the constants routed for now. I was focusing mainly on the baking and endorsing rights. Those constants in tezos_messages are most likely incorrect. We could also serve the constatns RPC from the new storage, but only the constants used by the rights RPCs are properly tested, so I wouldn't risk it now. Will create a jira shortly.

bkontur commented on rpc/src/services/protocol/proto_004/rights_service.rs:
```diff
@@ -93,22 +107,48 @@ pub(crate) fn get_baking_rights(
+                    (
+                        *roll,
+                        SignaturePublicKeyHash::from_tagged_bytes(delegate.clone())
+                            .unwrap()
```
@adonagy 
unwrap?

adonagy replied:
Oh, this is not good, will have to rework this

bkontur replied:
@adonagy 
maybe if you use `into_iter` , you dont need to `delegate.clone()`  ?

bkontur commented on rpc/src/services/protocol/proto_006/mod.rs:
```diff
@@ -1,6 +1,81 @@
+use super::{string_to_int, vec_string_to_int};
+
+#[derive(Debug, Deserialize, Clone, CopyGetters)]
+pub(crate) struct ProtocolConstants {
```
@adonagy 
are these `ProtocolConstants` structs different from those constants from `tezos_messages/src/protocol/proto_*`,
couldnt be better reused? is it possible?

adonagy replied:
Same reason as here: https://github.com/tezedge/tezedge/pull/782#discussion_r710857033

bkontur commented on rpc/tests/integration_tests.rs:
```diff
@@ -266,15 +552,15 @@ async fn integration_tests_rpc(from_block: i64, to_block: i64) {
-        let preserved_cycles = constants_json["preserved_cycles"]
-            .as_i64()
-            .unwrap_or_else(|| panic!("No constant 'preserved_cycles' for block_id: {}", level));
+        // let PRESERVED_CYCLES = constants_json["PRESERVED_CYCLES"]
```
@adonagy 
do we need this commented code?

bkontur commented on storage/src/constants_storage.rs:
```diff
@@ -0,0 +1,78 @@
+
+pub type ConstantsStorageKV = dyn TezedgeDatabaseWithIterator<ConstantsStorage> + Sync + Send;
+
+// TODO: double check the type
```
@adonagy 
TODO?

adonagy replied:
leftover TODO, already checked

bkontur commented on storage/src/cycle_storage.rs:
```diff
@@ -0,0 +1,109 @@
+// Copyright (c) SimpleStaking, Viable Systems and Tezedge Contributors
+// SPDX-License-Identifier: MIT
+
+// use std::io::SeekFrom;
```
@adonagy 
commented import?

adonagy replied:
I'll remove it

bkontur commented on storage/src/lib.rs:
```diff
@@ -310,6 +327,27 @@ pub fn store_applied_block_result(
+
+    // store new cycle eras if they are present
+    if let Some(new_cycle_eras) = block_result.new_cycle_eras_json {
+        println!(
```
@adonagy 
why `println!`? why not to move here `log: &Logger`?

adonagy replied:
Oh, just a dangling debug print that slipped trough my fingers, will remove

bkontur reviewed (changes requested):
@adonagy 
Ado, please, check comments, at least those about reuse cached services could improve performance little bit more

adonagy commented:
@bkontur Thank you for the review! Will address the comments now.
